File: python/src/agent_sdk/contracts.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, AsyncIterator, Awaitable, Callable, Mapping, Protocol

from .models import (
    ComputingSession,
    NetworkMessageAction,
    NetworkMessageType,
)


@dataclass(frozen=True, slots=True)
class RuntimeHttpResponse:
    status_code: int
    body: Mapping[str, Any]


class ProofVerifier(Protocol):
    async def verify_group_config(self, payload: Mapping[str, Any]) -> None: ...


class MessageSigner(Protocol):
    async def sign_a2a(self, payload: Mapping[str, Any]) -> Mapping[str, Any]: ...


class MessageSignatureVerifier(Protocol):
    async def verify_a2a(
        self, payload: Mapping[str, Any], expected_did_key: str
    ) -> None: ...


class ControlRequestAuthenticator(Protocol):
    async def authenticate(
        self, path: str, payload: Mapping[str, Any]
    ) -> Mapping[str, Any]: ...


class RuntimeTransport(Protocol):
    async def get_ue_info(self) -> Mapping[str, Any]: ...

    async def get_acn_status(self) -> Mapping[str, Any]: ...

    async def start_downlink(
        self,
        handler: Callable[
            [str, int, Mapping[str, Any]], Awaitable[Mapping[str, Any] | None]
        ],
        on_reconnected: Callable[[], Awaitable[None]] | None = None,
    ) -> None: ...

    async def request(
        self, method: str, path: str, body: Mapping[str, Any]
    ) -> Mapping[str, Any]: ...

    async def request_with_status(
        self, method: str, path: str, body: Mapping[str, Any]
    ) -> RuntimeHttpResponse: ...

    async def close(self) -> None: ...


class SandboxTransport(Protocol):
    """HTTP transport for C-02 Sandbox user-plane endpoints."""

    async def request_with_status(
        self,
        method: str,
        url: str,
        body: Mapping[str, Any] | None,
        timeout_seconds: float,
        source_ipv4: str,
    ) -> RuntimeHttpResponse: ...

    async def close(self) -> None: ...


class TunDevice(Protocol):
    name: str
    cidr: str
    mtu: int

    async def read(self) -> bytes: ...

    async def write(self, packet: bytes) -> None: ...

    async def close(self) -> None: ...


PacketHandler = Callable[[bytes], Awaitable[None]]


class ConnectIpTransport(Protocol):
    @property
    def connected(self) -> bool: ...

    async def start(self, on_packet: PacketHandler) -> None: ...

    async def send_packet(self, packet: bytes) -> None: ...

    async def close(self) -> None: ...


class LocalServer(Protocol):
    async def start(
        self,
        *,
        agent_ip: str,
        tcp_port: int,
        udp_port: int,
        on_a2a_message: Callable[[Mapping[str, Any]], Awaitable[None]],
    ) -> None: ...

    async def close(self) -> None: ...


class PeerMessenger(Protocol):
    async def send(
        self, endpoint: str, body: Mapping[str, Any], timeout: float
    ) -> Mapping[str, Any]: ...


class NetworkMessageListener(Protocol):
    async def on_network_message(
        self, message_type: NetworkMessageType, payload: Mapping[str, Any]
    ) -> NetworkMessageAction: ...


class GroupMessageListener(Protocol):
    async def on_group_message(
        self, group_id: str, sender_agent_id: str, payload: Mapping[str, Any]
    ) -> None: ...


class VideoUploadHandle(Protocol):
    track_id: str
    state: str

    async def pause(self) -> None: ...

    async def resume(self) -> None: ...

    async def stop(self) -> None: ...


class RemoteVideoStream(Protocol):
    def __aiter__(self) -> AsyncIterator[Any]: ...

    async def recv(self) -> Any: ...

    async def close(self) -> None: ...


class PreparedVideoUpload(Protocol):
    """A local send-only PeerConnection whose non-Trickle Offer is ready."""

    offer_sdp: str

    async def apply_answer(
        self, answer_sdp: str, timeout_seconds: float
    ) -> VideoUploadHandle: ...

    async def abort(self) -> None: ...


class PreparedProcessedVideo(Protocol):
    """A local receive-only PeerConnection whose non-Trickle Offer is ready."""

    offer_sdp: str

    async def apply_answer(
        self, answer_sdp: str, timeout_seconds: float
    ) -> RemoteVideoStream: ...

    async def abort(self) -> None: ...


class MediaOffloadAdapter(Protocol):
    """Platform WebRTC adapter; HTTP signaling is owned by AgentSdk."""

    def supports_video_codec(self, codec: str) -> bool: ...

    async def prepare_video_upload(
        self,
        session: ComputingSession,
        *,
        camera_id: int,
        width: int,
        height: int,
        fps: int,
        bitrate_kbps: int,
    ) -> PreparedVideoUpload: ...

    async def prepare_processed_video(
        self,
        session: ComputingSession,
    ) -> PreparedProcessedVideo: ...

    async def close(self) -> None: ...

File: android/agent-sdk/src/main/kotlin/com/rayneo/agent/sdk/transport/Contracts.kt
package com.rayneo.agent.sdk.transport

import com.rayneo.agent.sdk.model.NetworkMessageAction
import com.rayneo.agent.sdk.model.NetworkMessageType
import com.rayneo.agent.sdk.model.ComputingSession
import kotlinx.serialization.json.JsonObject
import java.net.URI

internal fun interface ProofVerifier {
    suspend fun verifyGroupConfig(payload: JsonObject)
}

internal fun interface ControlRequestAuthenticator {
    suspend fun authenticate(path: String, payload: JsonObject): JsonObject
}

internal interface DevicePublicKeyProvider {
    fun ensure()
    val publicKeyBase64: String
}

internal interface MessageSigner {
    suspend fun signA2a(payload: JsonObject): JsonObject
}

internal interface MessageSignatureVerifier {
    suspend fun verifyA2a(payload: JsonObject, expectedDidKey: String)
}

interface RuntimeTransport {
    suspend fun getUeInfo(): JsonObject
    suspend fun getAcnStatus(): JsonObject
    suspend fun startDownlink(
        onReconnected: suspend () -> Unit = {},
        handler: suspend (String, Int, JsonObject) -> JsonObject?,
    )
    suspend fun request(method: String, path: String, body: JsonObject): JsonObject
    suspend fun requestWithStatus(
        method: String,
        path: String,
        body: JsonObject,
    ): RuntimeHttpResponse = RuntimeHttpResponse(200, request(method, path, body))
    suspend fun requestWithStatus(
        method: String,
        path: String,
        body: JsonObject,
        timeoutSeconds: Double,
    ): RuntimeHttpResponse = requestWithStatus(method, path, body)
    suspend fun close()
}

internal interface SandboxTransport {
    suspend fun requestWithStatus(
        method: String,
        url: String,
        body: JsonObject?,
        timeoutSeconds: Double,
        sourceIpv4: String?,
    ): RuntimeHttpResponse
    suspend fun uploadWithStatus(
        url: String,
        fields: Map<String, String>,
        fileFieldName: String,
        fileName: String,
        contentType: String,
        content: ByteArray,
        timeoutSeconds: Double,
        sourceIpv4: String?,
    ): RuntimeHttpResponse
    suspend fun close()
}

data class RuntimeHttpResponse(
    val statusCode: Int,
    val body: JsonObject,
)

data class TunnelConfiguration(
    val agentTunCidr: String,
    val routes: Set<String>,
    val mtu: Int,
)

interface TunnelController {
    val tunFd: Int
    val clientIdentityDirectory: String
    suspend fun establish(configuration: TunnelConfiguration)
    suspend fun replaceGroupPeers(groupId: String, peerIps: Set<String>)
    fun currentAllowedPeerIps(): Set<String>
    fun setTunFdSwapper(swapper: suspend (Int) -> Unit)
    fun setTunReplacedListener(listener: suspend () -> Unit) = Unit
    suspend fun close()
}

data class MasqueConfiguration(
    val serverUrl: String,
    val authorization: String?,
    val localVlanIp: String,
    val agentTunCidr: String,
    val mtu: Int,
    val identityDirectory: String,
)

fun interface LocalAddressResolver {
    fun resolve(serverUri: URI): String
}

interface MasqueTransport {
    val connected: Boolean
    suspend fun start(tunFd: Int, configuration: MasqueConfiguration)
    suspend fun replaceTunFd(tunFd: Int)
    fun statistics(): MasqueTransportStatistics = MasqueTransportStatistics()
    suspend fun close()
}

data class MasqueTransportStatistics(
    val downlinkPackets: Long = 0,
    val downlinkPacketsOverTunMtu: Long = 0,
    val downlinkReadBufferTooSmall: Long = 0,
    val uplinkDatagramTooLarge: Long = 0,
    val maxDownlinkPacketBytes: Long = 0,
)

interface LocalServer {
    suspend fun start(
        agentIp: String,
        tcpPort: Int,
        udpPort: Int,
        onA2aMessage: suspend (JsonObject) -> Unit,
    )
    suspend fun close()
}

interface PeerMessenger {
    suspend fun send(endpoint: String, body: JsonObject, timeoutMillis: Long): JsonObject
}

fun interface NetworkMessageListener {
    suspend fun onNetworkMessage(
        messageType: NetworkMessageType,
        payload: JsonObject,
    ): NetworkMessageAction
}

fun interface GroupMessageListener {
    suspend fun onGroupMessage(
        groupId: String,
        senderAgentId: String,
        payload: JsonObject,
    )
}

interface VideoUploadHandle {
    val trackId: String
    val state: String
    suspend fun pause()
    suspend fun resume()
    suspend fun stop()
}

/** Wrapper around the platform WebRTC VideoTrack owned by a media adapter. */
interface VideoTrack {
    val trackId: String
    fun addSink(sink: Any)
    fun removeSink(sink: Any)
}

interface ProcessedVideoStream {
    val track: VideoTrack
    val state: String
    suspend fun close()
}

internal interface LocalProcessedVideo {
    val track: VideoTrack
    suspend fun close()
}

internal interface PreparedMediaConnection<T> {
    val offerSdp: String
    suspend fun applyAnswer(answerSdp: String, timeoutSeconds: Double): T
    suspend fun abort()
}

internal interface MediaOffloadAdapter {
    fun supportsVideoCodec(codec: String): Boolean

    suspend fun prepareVideoUpload(
        session: ComputingSession,
        cameraId: String,
        width: Int,
        height: Int,
        fps: Int,
        bitrateKbps: Int,
        timeoutSeconds: Double,
    ): PreparedMediaConnection<VideoUploadHandle>

    suspend fun prepareProcessedVideo(
        session: ComputingSession,
        timeoutSeconds: Double,
    ): PreparedMediaConnection<LocalProcessedVideo>

    suspend fun close()
}
